toolchain/lower/file_context.h

```diff
@@ -219,6 +308,28 @@ class FileContext {
 
   // Global format string for `printf.int.format` used by the PrintInt builtin.
   llvm::Value* printf_int_format_string_ = nullptr;
+
+  // For a generic function, keep track of the specifics for which LLVM
+  // function declarations were created. Those can be retrieved then via
+  // from specific_functions_, via specific_functions_[specific_id.index].
+  Map<SemIR::GenericId, llvm::SmallVector<SemIR::SpecificId>> lowered_specifics;
+  // For specifics that exist in lowered_specifics, store their function type
+  // information: return and parameter types.
+  // TODO: Storing all members of FunctionTypeInfo may not be necessary.
+  Map<SemIR::SpecificId, FunctionTypeInfo> lowered_specifics_types;
+
+  // This is initialized and populated while lowering a specific.
+  // The list of states for a specific is populated by the function_context
+  // for that specific.
+  Map<SemIR::SpecificId, llvm::SmallVector<LoweringFunctionState>>
+      lowered_states;
+
+  // Equivalent specifics found. TODO: add flipped pairs
+  Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>> equivalent_specifics;
```

Comment:
I think a [disjoint set data structure](https://en.wikipedia.org/wiki/Disjoint-set_data_structure) would be much more efficient (in both space and time), and automatically solve the flipped-pairs issue.

Reply:
Sorry, forgot to delete this comment when I moved it to reply to Jon's below.